kafka/admin/client.py

```diff
@@ -472,14 +472,48 @@ def delete_topics(self, topics, timeout_ms=None):
                 .format(version))
         return response
 
-    # list topics functionality is in ClusterMetadata
-    # Note: if implemented here, send the request to the least_loaded_node()
 
-    # describe topics functionality is in ClusterMetadata
-    # Note: if implemented here, send the request to the controller
+    def _get_cluster_metadata(self, topics=None, auto_topic_creation=False):
+        """
+        topics == None means "get all topics"
+        """
+        version = self._matching_api_version(MetadataRequest)
+        if version <= 3:
+            if auto_topic_creation:
+                raise IncompatibleBrokerVersion(
+                    "auto_topic_creation requires MetadataRequest >= v4, which"
+                    " is not supported by Kafka {}"
+                    .format(self.config['api_version']))
 
-    # describe cluster functionality is in ClusterMetadata
-    # Note: if implemented here, send the request to the least_loaded_node()
+            request = MetadataRequest[version](topics=topics)
+        elif version <= 5:
+            request = MetadataRequest[version](
+                topics=topics,
+                allow_auto_topic_creation=auto_topic_creation
+            )
+
+        future = self._send_request_to_node(
+            self._client.least_loaded_node(),
+            request
+        )
+        self._wait_for_futures([future])
+        return future.value
+
+    def list_topics(self):
+        metadata = self._get_cluster_metadata(topics=None)
+        obj = metadata.to_object()
+        return [t['topic'] for t in obj['topics']]
+
+    def describe_topics(self, topics=None):
+        metadata = self._get_cluster_metadata(topics=topics)
+        obj = metadata.to_object()
+        return obj['topics']
+
+    def describe_cluster(self):
+        metadata = self._get_cluster_metadata()
+        obj = metadata.to_object()
+        obj.pop('topics')  # We have 'describe_topics' for this
```

Comment:
Yeah, based on [their javadoc](https://www.apache.org/dist/kafka/2.3.0/javadoc/org/apache/kafka/clients/admin/DescribeClusterResult.html) they exclude topics as part of the metadata call, though it also looks like they [avoid fetching the topics at all](https://github.com/apache/kafka/blob/738e14edb833f3efdf807c506531b3eed4ee7137/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L1697).

It's easy enough to throw in a `def describe_cluster_and_topics` if you want to deviate from the java client a bit?

Reply:
Let's leave as-is if it follows Java, that will be good enough for now and anyone who wants the speed boost can live dangerously with `_get_cluster_metadata()`.

Also, my reading of that Java code is they're passing an empty list for the topics, which maps to fetching all topics... but I could be wrong it's been over a year since I looked at the Metadata protocol definition.

Reply:
It's been a bit for myself as well, I'm testing it now. I also just noticed that I had missed the "use controller for describe topics" comment, I'm adding that now

Reply:
Oh right, I missed that too!

Reply:
FTR this was fixed in https://github.com/dpkp/kafka-python/pull/1995